scvi/train/_metrics.py

```diff
@@ -45,29 +39,115 @@ def mode(self):
         self.n_obs += n_obs_minibatch
         self.n_batches += 1
 
     def compute(self):
-        avg_reconstruction_loss = self.reconstruction_loss / self.n_obs
-        avg_kl_local = self.kl_local / self.n_obs
-        avg_kl_global = self.kl_global / self.n_batches
-        # elbo on the scale of one observation
-        elbo = (
-            avg_reconstruction_loss + avg_kl_local + (avg_kl_global / self.n_obs_total)
-        )
-
-        return elbo
+        avg_elbo_component = self.elbo_component / self.n_obs
+
+        return avg_elbo_component
+
+
+class ReconstructionLossMetric(BaseElboMetric):
+    """
+    Reconstruction loss metric aggregator for scvi-tools experiments.
+
+    Parameters
+    ----------
+    mode
+        Train or validation, used for logging names
+    dist_sync_on_step
+        optional, by default False
+    **kwargs
+        Keyword args for :class:`torchmetrics.Metric`
+    """
+
+    def __init__(
+        self,
+        mode: Literal["train", "validation"],
+        dist_sync_on_step: bool = False,
+        **kwargs
+    ):
+        super().__init__(mode=mode, dist_sync_on_step=dist_sync_on_step, **kwargs)
+
+    def update(
+        self,
+        reconstruction_loss_sum: torch.Tensor,
+        n_obs_minibatch: int,
+    ):
+        """Updates all metrics."""
+        super().update(reconstruction_loss_sum, n_obs_minibatch)
+
+
+class KLLocalMetric(BaseElboMetric):
+    """
+    KL local loss metric aggregator for scvi-tools experiments.
+
+    Parameters
+    ----------
+    mode
+        Train or validation, used for logging names
+    dist_sync_on_step
+        optional, by default False
+    **kwargs
+        Keyword args for :class:`torchmetrics.Metric`
+    """
+
+    def __init__(
+        self,
+        mode: Literal["train", "validation"],
+        dist_sync_on_step: bool = False,
+        **kwargs
+    ):
+        super().__init__(mode=mode, dist_sync_on_step=dist_sync_on_step, **kwargs)
+
+    def update(
+        self,
+        kl_local_sum: torch.Tensor,
+        n_obs_minibatch: int,
+    ):
+        """Updates all metrics."""
+        super().update(kl_local_sum, n_obs_minibatch)
+
+
+class KLGlobalMetric(BaseElboMetric):
+    """
+    KL global loss metric aggregator for scvi-tools experiments.
+
+    Parameters
+    ----------
+    mode
+        Train or validation, used for logging names
+    dist_sync_on_step
+        optional, by default False
+    **kwargs
+        Keyword args for :class:`torchmetrics.Metric`
+    """
+
+    def __init__(
+        self,
+        mode: Literal["train", "validation"],
+        dist_sync_on_step: bool = False,
+        **kwargs
+    ):
+        super().__init__(mode=mode, dist_sync_on_step=dist_sync_on_step, **kwargs)
+
+    def update(
+        self,
+        kl_global: torch.Tensor,
+        n_obs_minibatch: int,
+    ):
+        """Updates all metrics."""
+        super().update(kl_global, n_obs_minibatch)
+
+    def compute(self):
```

Comment:
this looks like the only thing that sets apart this class from the rest. Otherwise a lot of copy paste. To avoid over generalizing, maybe you could just stick to having the base class, except it takes a metric name, and you add the kwarg for what to normalize by?